### ISSUE TITLE
10.0.0-RC1 - Cable management not working as expected

### ISSUE THREAD

**Issue by rbrannath** [glpi-project/glpi#10720]
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

10.0.0-RC1

### Bug description

First of all: I am not quite sure if this is a bug or intended behaviour or if this feature is even finished.
At the moment I assume, it is not finished and I would like to contribute. 

Unfortunately, I couldn't find any documentation of the new feature "cable management".
Therefore a few questions and thoughts about this feature:

1. In versions <= 9.5, there was a entity named "network outlet" (or networkpoint). This entity is gone now and I assume, its replacement is the new entity "socket". A couple of items can have a "socket" (e. g. computer, phones, printers), but this isn't what I expect from a simple "network outlet" which is just in the wall of a room. Also, there are "network ports" too, which can be assigned to a socket. What exactly is a "socket" supposed to be?

2. If a "socket" is basically just a connect point of a device (which can be RJ45, USB, HDMI, and so on), then I am missing the "network outlet" as a kind of device.

3. For example, there is a switch (network device), patchpanel (passive equipment) and a computer. The physical connection would be as follows:
```
         Switch (Port 1)
                   |  
         Patchpanel (Port 8)
                   |
       Networkoutlet (1.08) -> in room A014    
                   |
          Computer (Port 1)
```
How to map this common setup in GLPI?
The switch and the computer can have network ports, but not the patchpanel (passive device). Also, if using sockets, I would have to add a socket for each network port which would be a 1:1 relation.

4. The cable management feature could be used for any other connection type too (e. g. USB or HDMI). For this, the other connection types (network ports, ports of printer / monitors) should be deactivated.

My opinion is as follows:
An separate entity network outlet is still needed, which then can be connected to the front (computer, accesspoint) and to the back (patchpanel). The existing ports (of computer, printes, monitors, etc.) should be removed and generalized into sockets and those sockets should include more information (e. g. a socket type "RJ45" has a speed of 1 Gbit). A switch (network device) would not have 24 ports, but 24 sockets of type RJ45 and a monitor several sockets of the types HDMI, Displayport and so on. Those sockets may only be assigned once for consistency.

This would lead to a efficient use of connections between different devices and opens the possibility of a graphical network map (which is a feature request for sure :))

I look forward to the answers and opinions.

**Comment by rbrannath** [comment]
@cconard96 @cedric-anne could you give me a feedback on this? For me and many other institutions in the education sector, this matter is crucial for using GLPI, as we have to document the cables.

**Comment by orthagh** [comment]
@stonebuzz

**Comment by husisusi** [comment]
Hey guys,
it's a good question. any news?

**Comment by stonebuzz** [comment]
hi @rbrannath 

TLDR : for us, a network outlet is nothing more than a socket

# NetworkPort

From our point of view the ```NetworkPort``` object represents the virtual ports of a network card.

And contain (among other things) the following information:
* The port number
* The speed 
* The specialization (ethernet / fiber etc ...)
* MAC
* Network name
* IP addresses
 
# Socket

The ```Socket``` object represents the physical part.

And contain (among other things) the following information:
* Type
* Location (from old object ```netpoints```) 
* Position
    * usefull if device have no NetworkPort -> Passive Device
    * if NetworkPort is defined, the position is taken over  
* Wiring Side
* ```Asset``` / ```NetworkPort```


This object (```Socket```) can be linked directly to a ```NetworkPort``` from an ```Asset```

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b6835.png)


or only to an ```Asset``` without ```NetworkPort``` ie :```Passive Device```

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b6836.png)


In both cases, the ```Socket``` will be visible in the ```Asset```

From ```Asset``` with ```NetworkPort``` ie :```Comuter```
![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b6837.png)

From ```Asset``` without ```NetworkPort```  ie: ```Passive Device```

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b6838.png)

# NetworkOutlet

If I understand correctly you want to represent the network socket (```NetworkOutlet```) in room

```NetworkOutlet``` is nothing more than a ```Socket``` IMHO

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b6839.png)

The notion of ```Room``` in the ```office``` sense does not exist in GLPI. We have ```Server Room``` but nothing to represent an ```office```

Perhaps you use the GLPI locations to represent office ?

In this case, a ```Socket``` can be linked to a location only

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b683a.png)

It would then be enough to modify the ```Cable``` object to allow a link on a ```Socket``` from a GLPI ```Location```

**Comment by rbrannath** [comment]
Hi @stonebuzz ,

thank you very much for the detailed answers of my questions and explanation of the terms.
This really helped me to understand the concept of the cable management and what a network port means in GLPI. For me it was the physical port, not a virtual port.

It would be nice to have this documented somewhere. All I found about this topic is here and it is partly in french, but there it states it can be both -> https://glpi-project.org/DOC/EN/
I am not sure if these docs are outdated.

**Followup question:**
A socket (network outlet) in a room/location is connected to a computer/phone/printer, that's obvious. But what about a socket which is connected to a switch? In a server room there is no such thing as a network outlet (socket) but rather patchpanels, which contain multiple sockets. I can add a socket to the switch and connect this socket to the socket of the computer (this connection would represent the patchpanel, but it does not explicit document the patchpanel). Is this the way to go? I would rather have a passive device representing the patchpanel.

**Some thoughts about this:**
A socket is basically a network outlet, as you mentioned. In my opinion, a networkoutlet must have two connection points.
    1. is in a room/location -> this is where a computer is connected (this is the front)
    2. is in the server room / network cabinet / location -> this is where a switch is connected (this is the back)
I don't see a reason why sockets, who are physically connected and inseparable, should be managed separately in GLPI.
It would be nicer to have a patchpanel asset (or even the existing passive device), which then has numerous sockets with endpoints. Each endpoint could have a location and another asset type connected. Basically, an endpoint would be what a socket is right now.
Given the above, having a patchpanel with 24 Ports there would be 24 sockets, 48 socket endpoints which can be connected to 48 devices (e. g. computer/switch). Resulting in 48 managable elements.
With the current way, having a patchpanel which is a passive device,  with 24 front sockets + 24 back sockets. Those sockets are connected to 24 sockets (computers) and connected to 24 sockets (switch). Resulting in 96 managable elements for the same setup.

Another point is, having a building or in my case, multiple buildings and hundred or even thousands of sockets, there must be a bulk CRUD function like there is for network ports. There must be a handable solution for this.

I hope you understand, what my point is and I am looking forward for your opinions.

**Comment by stonebuzz** [comment]
Hi @rbrannath 

Currently it is possible to represent this path :

```NetworkEquipement``` -> ```PatchPanl``` -> ```Computer```

# Socket

Each asset has a ```Socket``` linked to a ```NetworkPort``` (or to itself directly).


## Computer Socket
![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b683b.png)

## PatchPanel Socket
![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b683c.png)

## Switch Socket
![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b683d.png)

# Cable

Actually only two cables are needed to get 

```NetworkEquipement``` -> ```PatchPanel``` -> ```Computer```

> We agree that the ```NetworkOutlet``` is missing


# Switch to patch panel

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b683e.png)


# PatchPanel to computer 

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b683f.png)


> Note : Is not mandatory to double the ```Socket``` on the patch panel
> Two different cables can use the same ```Socket```

![](https://hedgedoc.glpi-network.com/uploads/ee6850c6e06833fb6c90b6840.png)


# NetworkOutlet


If we want this path

```NetworkEquipement``` -> ```PatchPanel``` -> ```Location``` (```NetworkOutlet```)  -> ```Computer```


We must allow the connection of a ```Cable``` (```Endpoint A``` / ```Endpoint B```) to a GLPI location.


But I don't know if we want this kind of implementation (is generic? How the community currently uses the locations?)

Shouldn't we (maybe) create a ```Building``` / ```Office``` object in the ```Location```?

OR 

 create a ```NetworkOutlet``` linked to a  ```Location```?

Best regards

**Comment by rbrannath** [comment]
@cconard96 Thanks, I will have a look.
Maybe the old documentation should be hidden or at least marked as outdated.
This is the first result when google search for "glpi docs".

**Comment by macpacrac** [comment]
Please clarify how to use if the path

switch -> optical patch panel -> optical patch panel -> switch

If I connect switch -> optical patchpanel the cable is shown at the socket due to the binding to the network port of the switch.
If I link optical patchpanel -> optical patchpanel no cable is shown at the sockets (in the cable list it is shown).

**Comment by rbrannath** [comment]
@stonebuzz 

**Cable connection**
I agree on the way the connection is done in your example. This is how it can be done, but it is missing the network outlet, as you pointed out
> `NetworkEquipement `-> `PatchPanel `-> `Computer`

I would like to state herewith that a `NetworkOutlet`<> `Socket`. Can we agree on this?

**Location**
For me, a location object is a room (this can be a office, classroom, washing room, server room, or whatever). I do not distinguish between a server room or another type of room. This is because in my buildings, we rarely have rooms that are used exclusively for servers and network technology (which would be a datacenter in GLPI). Often the rooms are also used for other purposes.
It would be interesting to know, how other companies are solving this.

**NetworkOutlet**
As I pointed out in my first comment:
> An separate entity network outlet is still needed (...)

and you suggested

> create a NetworkOutlet linked to a Location?

I suggest to reenable the abandonded object `NetworkOutlet`. and make it linkable to a socket. As far as I know, there was already a link to a `Location`.

**Socket**
Another suggestion would be to get rid of `Location` in a socket. If a socket is the physical representation of the network port, it would get its location from the asset type it is connected to.

**Comment by rbrannath** [comment]
> Please clarify how to use if the path
> 
> switch -> optical patch panel -> optical patch panel -> switch

> 
> If I connect switch -> optical patchpanel the cable is shown at the socket due to the binding to the network port of the switch. If I link optical patchpanel -> optical patchpanel no cable is shown at the sockets (in the cable list it is shown).

@macpacrac
This is because the query expects 1 row to be returned to display the cable. If you have used the socket before to connect to a switch, the query returns 2 rows (cables) and nothing is displayed. You should see a PHP error if you enable debug mode.

Anyway, this is interesting and should be considered when making connecting multiple cables to a single socket.

**Comment by macpacrac** [comment]
@rbrannath 
Thanks for the info, I hadn't researched it that way until now.
For cable management I use Netbox and there it is solved in this style:
Active devices (switch, router, computer, ip phone, etc) have an interface, a console port and a power port.
These ports can be further plugged in according to their function. 
console port -> console server port, front and rear port
power port -> power outlet and power feed
interface -> interface, front port, rear port and circuit termination

Passive devices (patch panels, sockets, optical patch panels, etc.) these have front and rear ports.
These have a 1:1 (e.g. RJ45) or n:1 (fibre optic cables) connection between them.
The n:1 binding is meant e.g. 24 front ports are spliced to 1 rear port where a one fiber cable (24 optical fiber) is connected and routed to another location.

The devices can be connected by cables.
Active device <-> Active device (interface <-> interface)
Active device <-> Passive <-> Passive <-> Passive <-> Passive <-> Active device
(interface <-> front/rear <-> rear/front <-> front/rear <-> rear/front <-> interface)
Interface can also be connected to the rear port.
Then the route and the cable length can be displayed. A cable map display would also be useful.

I tried a GLPI to see if we could merge multiple sw in our company into one.

**Comment by armanduchi** [comment]
Thanks all of you.
We came from glpi -9.X and now with glpi 10 we miss https://github.com/PR-gh/patchpanel
We are considering to migrate to racktables or netbox,

unless a development focused on "patchpanels" witch rear-front properties.
 
All issues related to connections is rather confusing in glpi.

Imagine a Datacenter with 20 racks and more +1000 connections.

**Comment by armanduchi** [comment]
Very Interesting.
Is there any possibility to define multiple sockets at once for a passive device ?

Imagine we ha a passive device "patchpanel 48 ports".

The same way at "locations" we can add multiple sockets (from 1 to N),
so be able at "passive device" to do the same

And also keep in each socket ther ordering position

![image](https://user-images.githubusercontent.com/62557199/161289510-9397d2e2-bf7d-4b04-9607-40b199b7922a.png)

**Comment by armanduchi** [comment]
The same problem with a network device.
A typical network device with 48 ports,
How could we define 48 ordered sockets ?

And multiply this for N network devices. ( N > 200 at my University)

**Comment by stonebuzz** [comment]
> @stonebuzz
> 
> **Cable connection** I agree on the way the connection is done in your example. This is how it can be done, but it is missing the network outlet, as you pointed out
> 
> > `NetworkEquipement `-> `PatchPanel `-> `Computer`
> 
> I would like to state herewith that a `NetworkOutlet`<> `Socket`. Can we agree on this?

For us, a ```NetworkOutlet``` is nothing more than a ```Socket```

By adding the notion of ```Location``` here we should be able to do what you want

![image](https://user-images.githubusercontent.com/7335054/161504096-d8597c81-613f-4024-91c3-0675c696696f.png)



> 
> **Location** For me, a location object is a room (this can be a office, classroom, washing room, server room, or whatever). I do not distinguish between a server room or another type of room. This is because in my buildings, we rarely have rooms that are used exclusively for servers and network technology (which would be a datacenter in GLPI). Often the rooms are also used for other purposes. It would be interesting to know, how other companies are solving this.
> 
> **NetworkOutlet** As I pointed out in my first comment:
> 
> > An separate entity network outlet is still needed (...)
> 
> and you suggested
> 
> > create a NetworkOutlet linked to a Location?
> 
> I suggest to reenable the abandonded object `NetworkOutlet`. and make it linkable to a socket. As far as I know, there was already a link to a `Location`.

```NetworkOutlet``` has never existed in GLPI.

We originally had a ```NetPoint``` object which was just renamed to ```Socket``` (same basic characteristics but with additional information -> ```NetworkPort```, ```Model``` ....).

> 
> **Socket** Another suggestion would be to get rid of `Location` in a socket. If a socket is the physical representation of the network port, it would get its location by the asset it is connected to.

Unless we decide to use ```Socket``` as a ```NetworkOutlet```

Best regards

**Comment by stonebuzz** [comment]
> The same problem with a network device. A typical network device with 48 ports, How could we define 48 ordered sockets ?
> 
> And multiply this for N network devices. ( N > 200 at my University)

If I am right, this part has not changed since the renaming of  ```NetPoint``` to ```Socket```

Originally it was not possible to add massively ```NetPoints``` to an asset (passive or active)

It was only possible on the ```Location``` object.

This part can be improved

**Comment by rbrannath** [comment]
@stonebuzz thanks for making it clear. I have already learned that `Socket` used to be `NetPoint`.
As like others stated here, the new cable management feature is welcomed but needs improvements in usability.

> Originally it was not possible to add massively NetPoints to an asset (passive or active)

This is correct. Many have used https://github.com/PR-gh/patchpanel for this purpose. It has a feature to add multiple `Netpoints` at once. Since this plugin is not compatible with GLPI 10 anymore, there is a void to fill.

I have been working several days for a solution and found a way to work with the cable management. 
Also, I have figured out some improvements and did some coding to implement this.

Definition:
1. `Socket` is indeed the "network outlet/netpoint" one will find in a room, but also the physical connection point of computers, phones, switches etc.
2. `PassiveDCEqupiments` are patchpanels, fiberpanels, network outlets/netpoints

Improvments:
1. Make it possible to add multiple sockets for an asset
![grafik](https://user-images.githubusercontent.com/14164020/161505888-a68ca4a3-dab1-44c0-a2aa-fb8aed48005c.png)
In the form of the new `Socket` are additional fields to make it easy to customize the name.
![grafik](https://user-images.githubusercontent.com/14164020/161509300-e9696e1f-ffca-40a8-86f6-d6cb861d4a32.png)

2. Split up the network port form, because it is confusing when using passive devices
![grafik](https://user-images.githubusercontent.com/14164020/161509539-53efc823-fb63-4548-ab5a-7096069447dc.png)

3. Make it easier to add/edit connected cables to a socket by adding connection info inside the socket form. Also, make it static to have a front and a rear connection. (In my opinion "front" means accessible by just plugin in, while "rear" means it is hidden, e. .g. a computer would just use the front connection and a passive device, like a patchpanel would use both).
By clicking "add" the socket and its connected asset will be put to "Endpoint A" in the cable form.
![grafik](https://user-images.githubusercontent.com/14164020/161511325-2c7d743c-4856-4a0f-947e-7da0203e9090.png)

5. Remove "wiring_side" from `Socket` and put it to `Cable`. If I am right, there is either a front <-> front or <-> rear <-> rear connection, so it would be redundant and lead to inconsistency if it would remain at the socket. The "wiring_side" inside the `Cable` determines the side of connection in the `Socket` form.

@ anybody
Please have a look at my branch for further testing: https://github.com/rbrannath/glpi/tree/10.0.0-RC3/cable_management_improvements and give me some feedback.

I will open an PR, as soon as I am fully done. If it is not desired, I will develop a plugin

**Comment by stonebuzz** [comment]
Hi @rbrannath 

I started looking at your branch, can you create a PR to discuss on a dedicated thread ?

Thank you again for your investment.

Best regards

**Comment by cedric-anne** [comment]
> Hi @rbrannath
> 
> I started looking at your branch, can you create a PR to discuss on a dedicated thread ?
> 
> Thank you again for your investment.
> 
> Best regards

I created it: https://github.com/glpi-project/glpi/pull/11186

**Comment by KongNan** [comment]
For cable management, other friends have previously encountered some problems (for example,  [on the forum](https://forum.glpi-project.org/viewtopic.php?id=35407))

Thanks for all the work, I'm eager to use it in GLPI10!

**Comment by KongNan** [comment]
Yes, a lot of `PatchPanels`!! I couldn't agree more

**Comment by Megachip** [comment]
> Yes, a lot of `PatchPanels`!! I couldn't agree more

^^ tried to follow #11186 but still not sure, will it possible to bind sockets between patch panels in 10.0.1.0?

**Comment by kmpm** [comment]
This issue should really be reopened.

Some references to suggestions and some of them are several years old.
* https://glpi.userecho.com/en/communities/1/topics/541-add-patch-panel-in-network-assets
* https://glpi.userecho.com/en/communities/1/topics/1786-it-would-be-useful-to-manage-physical-connections-in-racks-server-rooms-those-that-nicely-in
* https://glpi.userecho.com/en/communities/1/topics/979-improve-rack-management
* https://glpi.userecho.com/en/communities/1/topics/940-passive-painel-to-rack

Topic 940 refers to a patch that relates to passive equipment but that doesn't solve everything.
* https://github.com/glpi-project/glpi/pull/7118

**Comment by cedric-anne** [comment]
> This issue should really be reopened.

We only track bugs here.

The right place to talk about feature suggestions is https://suggest.glpi-project.org/.
The right place to talk about technical implementation it to [contact GLPI project team directly](https://glpi-project.org/contact/). You can also contact us if you want to sponsor some enhancements.
